src/lib.rs

```diff
@@ -144,20 +143,66 @@ pub fn run(args: Args) {
         (None, None) => ALL_NODES_QUERY.to_owned(),
         _ => unreachable!(),
     };
-    let specified_supported_language = args.language.map(|language| language.get_language());
     let cached_queries: CachedQueries = Default::default();
     let capture_index = CaptureIndex::default();
     let buffer_writer = BufferWriter::stdout(ColorChoice::Never);
     let matched = AtomicBool::new(false);
     let searched = AtomicBool::new(false);
 
-    get_project_file_parallel_iterator(specified_supported_language.as_deref(), &args.use_paths())
-        .for_each(|project_file_dir_entry| {
+    args.get_project_file_parallel_iterator().for_each(
+        |(project_file_dir_entry, matched_languages)| {
             searched.store(true, Ordering::SeqCst);
-            let language = maybe_supported_language_from_path(project_file_dir_entry.path())
-                .expect("Walker should've been pre-filtered to just supported file types");
+            if matched_languages.is_empty() {
+                match args.language() {
+                    Some(language) => {
+                        error_explicit_path_argument_not_of_specified_type(
+                            &project_file_dir_entry,
+                            language,
+                        );
+                    }
+                    None => {
+                        error_explicit_path_argument_not_of_known_type(&project_file_dir_entry);
+                    }
+                }
+                return;
+            }
+            let language = return_if_none!(if matched_languages.len() > 1
```

Comment:
I wonder if `return_if_none` could (optionally?) support using braces instead of parentheses? I was a little confused about the syntax here, because I was really not expecting ~30 lines inside of the parentheses.

Reply:
Ok actually you get that "for free" (you can invoke any declarative macro with any "brace type") afaik, and that's a nice idea here (to use `return_if_none! { ... }` presumably)

But per my comment the way this was written was questionable so I restructured the logic here (+ the preceding `if matched_languages.is_empty()` "special case") into more of a nested-`match` style which got rid of the need for `return_if_none!()` at all here

And actually by rewriting it became clearer (if I did want to still be silly about avoiding allocating n more small `Vec`'s) that in fact it was easy/natural to not do that and only allocate that extra `Vec` in the case that there are multiple matching languages (in auto-language mode) (at which point allocating that `Vec` makes the code for that case much simpler)

Reply:
Oh that's interesting. So that's true even for things like `vec![]`?

Reply:
Yup it's just [by convention](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c3a1c885281ba6ce107cb891376fd29f) that we always use `[...]` there